integrations/stripe/subscriptions/server/node/src/email.ts

```diff
@@ -0,0 +1,317 @@
+            email,
+        )
+    }
+
+    static async sendInvoicePaymentFailedWallet(
+        email: string,
+        data: {
+            charge: Charge
+        },
+    ) {
+        const { charge } = data
+
+        // Attempted Charges will always have a last_attempt_failure_reason
+        // and next_attempt_at
+        if (!charge.last_attempt_failure_reason) return
+        if (!charge.next_attempt_at) return
+
+        const address = charge.source_address
+        // This is just to show a short version of the address in the email
+        // E.g. 0x1234...7890
+        const shortAddress = this.shortAddress(address)
+
+        // When 'address_blacklisted_by_usdc' is returned, the charge is
+        // not retried. The charge.failed event is sent instead.
+        const template =
+            charge.last_attempt_failure_reason ===
+            'insufficient_spending_allowance'
+                ? Template.InvoicePaymentFailedWalletAllowance
+                : Template.InvoicePaymentFailedWalletBalance
+
+        this.sendEmail(
+            {
+                name: template,
+                data: {
+                    amount: charge.amount.toString(),
+                    token: charge.token,
+                    chain: charge.chain,
+                    address: shortAddress,
+                    next_attempt_at: formatDistanceToNow(
+                        charge.next_attempt_at * 1000,
+                    ),
+                },
+            },
+            email,
+        )
+    }
+
+    static async sendSubscriptionTrialWillEnd(
+        email: string,
+        data: {
+            subscription: Stripe.Subscription
+        },
+    ) {
+        const { subscription } = data
+
+        const trialEnd =
+            subscription.trial_end !== null
+                ? formatDistanceToNow(subscription.trial_end * 1000)
+                : null
+
+        this.sendEmail(
+            {
+                name: Template.SubscriptionTrialWillEnd,
+                data: {
+                    trial_end: trialEnd ?? undefined,
+                },
+            },
+            email,
+        )
+    }
+
+    private static sendEmail<T extends Template>(
+        template: TemplateData[T] extends undefined
+            ? { name: T; data?: TemplateData[T] }
+            : { name: T; data: TemplateData[T] },
+        toAddress: string,
+    ): void {
+        const emailPayload = {
+            Destination: {
+                ToAddresses: [toAddress],
+            },
+            Template: template.name,
+            TemplateData: {
+                ...template.data,
+                ...this.BASE_CONFIGURATION,
+            },
+            ReplyToAddresses: [
+                `${this.BASE_CONFIGURATION.organisation_name} <contact@example.com>`,
+            ],
+            Source: `${this.BASE_CONFIGURATION.organisation_name} <billing@example.com>`,
+        }
+        console.log('Sending email', JSON.stringify(emailPayload, null, 2))
```

Comment:
Do we want to leave this here, or should we remove it?

Reply:
I would argue we could just leave it here, so in case of running the demo, you can see the potential emails being delivered.